client-misk-spanner/src/main/kotlin/app/cash/backfila/client/misk/spanner/internal/SpannerBackfillOperator.kt

```diff
@@ -0,0 +1,108 @@
+package app.cash.backfila.client.misk.spanner.internal
+
+import app.cash.backfila.client.misk.spanner.SpannerBackfill
+import app.cash.backfila.client.spi.BackfillOperator
+import app.cash.backfila.client.spi.BackfilaParametersOperator
+import app.cash.backfila.protos.clientservice.GetNextBatchRangeRequest
+import app.cash.backfila.protos.clientservice.GetNextBatchRangeResponse
+import app.cash.backfila.protos.clientservice.PrepareBackfillRequest
+import app.cash.backfila.protos.clientservice.PrepareBackfillResponse
+import app.cash.backfila.protos.clientservice.RunBatchRequest
+import app.cash.backfila.protos.clientservice.RunBatchResponse
+import com.google.cloud.spanner.Key
+import com.google.cloud.spanner.KeyRange
+import com.google.cloud.spanner.KeySet
+import com.squareup.moshi.Moshi
+import misk.moshi.adapter
+import okio.ByteString
+import okio.ByteString.Companion.encodeUtf8
+
+class SpannerBackfillOperator<Param : Any> internal constructor(
+  override val backfill: SpannerBackfill<Param>,
+  private val parametersOperator: BackfilaParametersOperator<Param>,
+  backend: SpannerBackend,
+) : BackfillOperator {
+  private var moshi: Moshi = backend.moshi
+  private val adapter = moshi.adapter<List<String>>()
+
+  override fun name() = backfill.javaClass.toString()
+
+  override fun prepareBackfill(request: PrepareBackfillRequest): PrepareBackfillResponse {
+    val config =
+      parametersOperator.constructBackfillConfig(request.parameters, request.dry_run)
+    backfill.validate(config)
+
+    val partitions = listOf(
+      PrepareBackfillResponse.Partition.Builder()
+        .backfill_range(request.range)
```

Comment:
are you requiring range to be passed in? In other implementations we compute the ranges if you don't pass it in

Reply:
The range is actually completely ignored. Spanner is unlike many other DBs, where for optimal performance primary keys really can't be in anything like a monotonic increasing range. I don't know how to compute a range without doing a full table scan, which seems... suboptimal.

Reply:
you can't ask for min/max primary key value?

Reply:
Primary keys are often random values like UUIDs and unordered for optimal performance. Min/max aren't valid concepts, as far as I can tell.

Reply:
Backfila requires ordered key values to operate. I'm curious how you would use it if that's not the case. I haven't used spanner but my understanding was its ordered, you just want to avoid sequential writes

Reply:
So the keys are ordered - but other keys could be added inside your key range during the backfill since they aren’t guaranteed to be an auto-incrementing integer.

From the spanner docs: “ Cloud Spanner stores rows in sorted order by primary key values”.

The reason you want well-distributed keys is because it helps prevent “hot shards” in a distributed system, which it has in common with dynamo. Therefore, I believe this should be fine.

Reply:
And to answer the original question - we don’t require a range to be passed in. That’s optional.

Reply:
Yeah I'm well aware how primary key design works in spanner, and you can have items added within the range. That's true even in auto increment, technically. It doesn't matter since the expectation is you are inserting new items that don't need backfilling.

It sounds like you are able to just ask spanner for records and it will give in some order, that should be fine I guess

Reply:
Wouldn't this work like dynamo backfills? Dynamo is somewhat different, but has a scan mechanism we use, and I believe we don't do ranges on it either? You could check that.

Reply:
Does this mean that you will essentially run your backfill single threaded?